### PR TITLE
sync: adds Notify for basic task notification

### PR DESCRIPTION
**Pull request by carllerche**
`Notify` provides a synchronization primitive similar to thread park /
unpark, except for tasks. Roughly, it is like a semaphore that starts with
no permits and `notify_one` adds a permit if, and only if, there currently are
no permits.

This effectively provides a basic way to get a task to wait until a signal is
received.

This PR also is a first step towards bringing the techniques used
by [futures-intrusive](https://github.com/Matthias247/futures-intrusive)
into Tokio. This introduces an intrusive linked list guarded by a mutex. Wakers are
stored on the (async) stack thanks to `Pin`. The details of the linked list are pulled
mostly from #2153, but with unused code removed, some slight API tweaks made,
and additional tests in the form of fuzzing added.

Because of the similarities between `Notify` and `Semaphore`, a question would be
whether `Notify` should exist at all. I am leaning towards having both `Notify
 and `Semaphore` as `Semaphore` does not include the necessary APIs to implement
`Notify`. Specifically, there is no `add_permit_if_none` function. Eventually, `Notify`
could be refactored to use `Semaphore` under the hood.

### CODE REVIEW

**Comment by carllerche** [comment]
As an example (included in api docs), this is how one could implement an MPSC channel using a `VecDeque` and `Notify`:

```rust
use tokio::sync::Notify;
use std::collections::VecDeque;
use std::sync::Mutex;

struct Channel<T> {
    values: Mutex<VecDeque<T>>,
    notify: Notify,
}

impl<T> Channel<T> {
    pub fn send(&self, value: T) {
        self.values.lock().unwrap()
            .push_back(value);

        // Notify the consumer a value is available
        self.notify.notify_one();
    }

    pub async fn recv(&self) -> T {
        loop {
            // Drain values
            if let Some(value) = self.values.lock().unwrap().pop_front() {
                return value;
            }

            // Wait for values to be available
            self.notify.recv().await;
        }
    }
}
```

**Comment by carllerche** [comment]
Another open question is what should the "receive" function be called. Right now, I have `Notify::recv`, but maybe it should be `Notify::acquire` to mirror `Semaphore` or maybe it should be something completely different.

File: tokio/src/sync/notify.rs [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+
+        // Safety: The type only transitions to a "Waiting" state when pinned.
+        let (notify, state, mut waiter) = unsafe { Pin::new_unchecked(self).project() };
+
+        // This is where we ensure safety. The `RecvFuture` is being dropped,
+        // which means we must ensure that the waiter entry is no longer stored
+        // in the linked list.
+        if let Waiting = *state {
+            let mut notify_state = WAITING;
+            let mut waiters = notify.waiters.lock().unwrap();
```
Comment (hawkw) [inline tokio/src/sync/notify.rs]: If the mutex is poisoned, this will double panic. But, if my reading of the comment above is correct, if we fail to remove the waiter from the list, we will have corrupted the list by leaving a pointer dangling. Should we instead "force" the mutex to unlock even if it is poisoned (e.g. by using the guard returned in a `PoisonError`), so that we can continue unwinding without leaving a dangling entry in the list?

File: tokio/src/util/linked_list.rs [hunk 2]
```diff
@@ -0,0 +1,483 @@
+//! An intrusive double linked list of data
+//!
+//! The data structure supports tracking pinned nodes. Most of the data
+//! structure's APIs are `unsafe` as they require the caller to ensure the
+//! specified node is actually
```
Comment (hawkw) [inline tokio/src/util/linked_list.rs]: ...is actually _what_? :)

File: tokio/src/util/linked_list.rs [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    /// self`.
+    pub(crate) fn pop_back(&mut self) -> Option<Pin<&mut T>> {
+        unsafe {
+            let mut last = match self.tail {
+                Some(tail) => {
+                    self.tail = tail.as_ref().prev;
+                    tail
+                }
+                None => return None,
+            };
```
Comment (hawkw) [inline tokio/src/util/linked_list.rs]: I think this could be simplified to
```rust
let mut last = self.tail?;
self.tail = last.as_ref().prev;
```

**Comment by hawkw** [comment]
> Another open question is what should the "receive" function be called. Right now, I have `Notify::recv`, but maybe it should be `Notify::acquire` to mirror `Semaphore` or maybe it should be something completely different.

How about `Notify::wait`? IMHO this best describes what the operation is actually trying to do: you are _waiting_ to be notified, not _receiving_ any actual data. That way, you have `Notify::wait` and `Notify::notify_one`, which kinda mirrors `std::sync::Barrier`.

File: tokio/src/sync/notify.rs [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+/// async fn main() {
+///     let notify = Arc::new(Notify::new());
+///     let notify2 = notify.clone();
+///
+///     tokio::spawn(async move {
+///         notify2.recv().await;
+///         println!("received a notification");
+///     });
+///
+///     notify.notify_one();
```
Comment (jonhoo) [inline tokio/src/sync/notify.rs]: Maybe include a `println!("sent notification")` before here, and say that they _will_ print in order.

File: tokio/src/sync/notify.rs [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+    ///
+    ///     notify.notify_one();
+    /// }
+    /// ```
+    pub fn notify_one(&self) {
+        // Load the current state
+        let mut curr = self.state.load(SeqCst);
+
+        // If the state is `EMPTY`, transition to `NOTIFIED` and return.
+        while let EMPTY | NOTIFIED = curr {
```
[the remark this answers is not in the document]
Comment (jonhoo) [reply tokio/src/sync/notify.rs]: I don't think that's the same? The original will match if `curr` is _either_ `EMPTY` _or_ `NOTIFIED`, whereas your will match if `curr` is equal to the binary or of `EMPTY` and `NOTIFIED`. At least unless I got my precedence rules messed up, the `|` in the `let` version will be interpreted as a pattern separator, not bitwise-or.

File: tokio/src/sync/notify.rs [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+            // The compare-exchange from `NOTIFIED` -> `NOTIFIED` is
+            // intended a happens-before synchronization must happen
```
Comment (jonhoo) [inline tokio/src/sync/notify.rs]: Full stop missing?

File: tokio/src/sync/notify.rs [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+) -> Option<Waker> {
+    loop {
+        match curr {
+            EMPTY | NOTIFIED => {
+                let res = state.compare_exchange(curr, NOTIFIED, SeqCst, SeqCst);
+
+                match res {
+                    Ok(_) => return None,
+                    Err(actual) => {
+                        curr = actual;
```
Comment (jonhoo) [inline tokio/src/sync/notify.rs]: Unless I'm reading this code wrong, `actual` _must_ be either `EMPTY` or `NOTIFIED` here. It cannot be `WAITING`, because that is only ever entered under the lock. Since we are going to continue until we succeed here anyway, and we are overwriting with a constant value, is there a reason to to just do a `state.store(NOTIFIED)`, `return None`, and call it a day?

File: tokio/src/sync/notify.rs [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+                    // Transition the state to WAITING.
+                    loop {
+                        match curr {
+                            EMPTY => {
+                                // Transition to WAITING
+                                let res = notify
+                                    .state
+                                    .compare_exchange(EMPTY, WAITING, SeqCst, SeqCst);
+
+                                if let Err(actual) = res {
```
Comment (jonhoo) [inline tokio/src/sync/notify.rs]: I believe `actual` _must_ be `NOTIFIED` here. Should we assert that?

File: tokio/src/sync/notify.rs [hunk 9]
```diff
@@ -0,0 +1,496 @@
...
+                                    .compare_exchange(NOTIFIED, EMPTY, SeqCst, SeqCst);
+
+                                match res {
+                                    Ok(_) => {
+                                        // Acquired the notification
+                                        *state = Done;
+                                        continue 'outer;
+                                    }
+                                    Err(actual) => {
+                                        curr = actual;
```
Comment (jonhoo) [inline tokio/src/sync/notify.rs]: I believe `actual` _must_ be `EMPTY` here. Should we assert that?

File: tokio/src/sync/notify.rs [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+}
+
+/// Initial "idle" state
+const EMPTY: u8 = 0;
+
+/// One or more threads are currently waiting to be notified.
+const WAITING: u8 = 1;
+
+/// Pending notification
+const NOTIFIED: u8 = 2;
```
Comment (jonhoo) [inline tokio/src/sync/notify.rs]: Should these be
```rust
#[repr(u8)]
enum State {
    Empty = 0,
    Waiting = 1,
    Notified = 2,
}
```
So that we avoid the
```rust
_ => unreachable!()
```
match arms?

File: tokio/src/sync/notify.rs [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+                                        curr = actual;
+                                        continue;
+                                    }
+                                }
+                            }
+                            _ => unreachable!(),
+                        }
+
+                        // The state has been transitioned to waiting
+                        break;
```
Comment (jonhoo) [inline tokio/src/sync/notify.rs]: This inverted flow is real weird. It'd be easier to read this (I think) if it was inverted so that the relevant cases used `break`, and the `continue` cases just naturally continued through the loop.

File: tokio/src/sync/notify.rs [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+                            w.waker = None;
+                            w.notified = false;
+
+                            // Remove the entry from the linked list. The entry
+                            // may only be in the linked list while the state is
+                            // `Waiting`.
+                            //
+                            // Safety: called while the lock is held
+                            unsafe {
+                                waiters.remove(waiter.as_mut());
```
Comment (jonhoo) [inline tokio/src/sync/notify.rs]: I'm confused -- isn't the `waiter` removed by the call to `pop_back` in the notify code? Why does it _also_ need to be removed here?

File: tokio/src/sync/notify.rs [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+        let (notify, state, mut waiter) = unsafe { Pin::new_unchecked(self).project() };
+
+        // This is where we ensure safety. The `RecvFuture` is being dropped,
+        // which means we must ensure that the waiter entry is no longer stored
+        // in the linked list.
+        if let Waiting = *state {
+            let mut notify_state = WAITING;
+            let mut waiters = notify.waiters.lock().unwrap();
+
+            // remove the entry from the list
```
Comment (jonhoo) [inline tokio/src/sync/notify.rs]: I'm surprised we don't have to re-check that the state is still `WAITING` here. What if just before we took the lock, a notify happened which notified us, removed us from the list, and set the state to `EMPTY`?

File: tokio/src/sync/notify.rs (hunk 10 above)
Comment (hawkw) [reply tokio/src/sync/notify.rs]: In order to turn the `u8` _into_ a `State`, there basically still needs to be a fallible match against a u8 _somewhere_

File: tokio/src/macros/pin.rs [hunk 14]
```diff
@@ -108,5 +108,13 @@ macro_rules! pin {
...
-    )* }
+    )* };
+    ($(
+            let $x:ident = $init:expr;
+    )*) => {
+        $(
+            let $x = $init;
+            crate::pin!($x);
+        )*
+    }
```
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/macros/pin.rs]: Ah yes, I meant to extract this to a separate PR.

File: tokio/src/sync/notify.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/sync/notify.rs]: because clippy?

File: tokio/src/sync/notify.rs [hunk 15]
```diff
@@ -0,0 +1,496 @@
...
+    ///         notify2.recv().await;
+    ///         println!("received a notification");
+    ///     });
+    ///
+    ///     notify.notify_one();
+    /// }
+    /// ```
+    pub fn notify_one(&self) {
+        // Load the current state
+        let mut curr = self.state.load(SeqCst);
```
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/sync/notify.rs]: There is no particular reason.

File: tokio/src/sync/notify.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Matthias247) [reply tokio/src/sync/notify.rs]: I think @jonhoo is right. The equivalent would be:

```rust
while curr == EMPTY || curr == NOTIFIED
```` 

It's probalby a bit more natural since an or pattern is not something one sees often (I haven't seen it before). But all versions should be ok.

File: tokio/src/sync/notify.rs (hunk 1 above)
Comment (Matthias247) [reply tokio/src/sync/notify.rs]: How would you expect the mutex to be able to become poisoned without being in a state that is so undefined that it doesn't matter anymore how we try to handle it?

File: tokio/src/sync/notify.rs [hunk 16]
```diff
@@ -0,0 +1,496 @@
...
+                        *state = Done;
+                        continue 'outer;
+                    }
+
+                    // Acquire the lock and attempt to transition to the waiting
+                    // state.
+                    let mut waiters = notify.waiters.lock().unwrap();
+
+                    // Reload the state with the lock held
+                    let mut curr = notify.state.load(SeqCst);
```
Comment (Matthias247) [inline tokio/src/sync/notify.rs]: Have you thought about just storing the `Waker` in the list, unlock it, and then check `state` again and act on a missed notification instead of trying to reason about the Mutex and the atomic at the same time?

I have the feeling it might be slightly less efficient for the case where the race actually happens, but might be easier to follow.

File: tokio/src/sync/notify.rs [hunk 17]
```diff
@@ -0,0 +1,496 @@
...
+/// ```
+///
+/// Unbound mpsc channel.
+///
+/// ```
+/// use tokio::sync::Notify;
+/// use std::collections::VecDeque;
+/// use std::sync::Mutex;
+///
+/// struct Channel<T> {
```
Comment (Matthias247) [inline tokio/src/sync/notify.rs]: I love example! However it also raises a question: How can we use `Notify` to implement traits - e.g. a channel that is composed of `Stream` and `Sink`. It seems like those would need to require having a `poll` fn exposed instead of an `async fn`.

I actually like the `async fn` much more since we should somehow try to get away from using bare `Future`s  and poll functions. But the composability story with `async fn`s is unfortunately still not there.

File: tokio/src/sync/notify.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hawkw) [reply tokio/src/sync/notify.rs]: @jonhoo oh, whoa, you're right — I forgot that `|` has a different meaning in patterns & interpreted it as binary or. IMO we should probably change this to 
```rust
while curr == EMPTY || curr == NOTIFIED
```
so that it's not misinterpreted by readers as `EMPTY.bit_or(NOTIFIED)`...

File: tokio/src/sync/notify.rs (hunk 1 above)
Comment (hawkw) [reply tokio/src/sync/notify.rs]: Hmm, you're right, I thought there were a couple cases where we made assertions while the lock is held, but after taking another look, that doesn't seem to be the case. Still, I'd prefer to err on the side of resilience if future changes add possible panics while holding a lock...

File: tokio/src/macros/pin.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/macros/pin.rs]: https://github.com/tokio-rs/tokio/pull/2274

File: tokio/src/sync/notify.rs (hunk 7 above)
Comment (carllerche) [reply tokio/src/sync/notify.rs]: `actual` can be `WAITING` at this point. A thread may concurrently store `WAITING` right before this thread acquires the lock.

Comment (carllerche) [reply tokio/src/sync/notify.rs]: Actually, the state is reloaded with the lock held, so I think you are right.

File: tokio/src/sync/notify.rs (hunk 13 above)
Comment (carllerche) [reply tokio/src/sync/notify.rs]: It took me a moment to remember, but I'm pretty sure it is correct. I added comments explaining.

File: tokio/src/sync/notify.rs (hunk 16 above)
Comment (carllerche) [reply tokio/src/sync/notify.rs]: It might work... not sure. I  mostly  modeled it after the park / unpark impl from `std`. Right now, it is roughly "check fast path, otherwise, lock and do the full work". This feels simpler than trying to "undo" work, but I can't say for sure.

 If you think it would be worth it, I can give a shot. wdyt?

File: tokio/src/sync/notify.rs (hunk 17 above)
Comment (carllerche) [reply tokio/src/sync/notify.rs]: I'm... punting on that 😆 I'm placing the emphasis on `async fn` for this primitive. For `poll` implementations, there is `AtomicWaker`, which is more or less the same thing but with wakers...

Do you have a suggestion on how to move forward?

File: tokio/src/sync/notify.rs (hunk 10 above)
Comment (carllerche) [reply tokio/src/sync/notify.rs]: I tried, it seems of limited value and makes the code messier (more conversions).

https://github.com/tokio-rs/tokio/commit/7d38153bf235299286052010742eb6dd1511213f

I'm leaning towards no. Thoughts?

File: tokio/src/sync/notify.rs (hunk 16 above)
Comment (Matthias247) [reply tokio/src/sync/notify.rs]: I have no strict preference on this one. Whatever seems to work and is understood by reviewers is fine. I thought an approach where you retry might be easier because of less require extra synchronization and code paths. The code for checking for a notification is already there, and you might just need to use it again. It's kind of the compare-and-swap approach we generally use for atomics.

File: tokio/src/sync/notify.rs (hunk 17 above)
Comment (Matthias247) [reply tokio/src/sync/notify.rs]: I think one thing we need to find out for this is where it will be used 😀
Some APIs like the one in the example will definitely do well with waiting just bein an `async fn`. It's hard to say whether it will also be used for poll based stuff, or whether implementations of those traits will become more rare anyway.

I guess it might be fair to include it as proposed, and do API modifications on a future pre 1.0 release if deemed necessary.

Comment (carllerche) [reply tokio/src/sync/notify.rs]: I like punting :)

File: tokio/src/sync/notify.rs (hunk 16 above)
Comment (carllerche) [reply tokio/src/sync/notify.rs]: I will opt to leave it this way as it follows the style of `park` / `unpark`... and i prefer not changing code if it isn't requested 😆

File: tokio/src/macros/pin.rs [hunk 18]
```diff
@@ -140,5 +140,5 @@ macro_rules! pin {
             let $x = $init;
             crate::pin!($x);
         )*
-    };
+    }
```
Comment (hawkw) [inline tokio/src/macros/pin.rs]: nit: this looks like an unrelated change?

File: tokio/src/sync/notify.rs [hunk 19]
```diff
@@ -0,0 +1,519 @@
...
+use std::task::{Context, Poll, Waker};
+
+/// Notify a single task to wake up.
+///
+/// `Notify` provides a basic mechanism to notify a single task of an event.
+/// `Notify` itself does not carry any data. Instead, it is to be used to signal
+/// another task to perform an operation.
+///
+/// `Notify` can be thought of as a [`Semaphore`] starting with 0 permits.
+/// [`notfied().await`] waits for a permit to become available and [`notify()`]
```
Comment (hawkw) [inline tokio/src/sync/notify.rs]: nits:
```suggestion
/// [`notified().await`] waits for a permit to become available, and [`notify()`]
```

File: tokio/src/sync/notify.rs [hunk 20]
```diff
@@ -0,0 +1,519 @@
...
+            waiters: Mutex::new(LinkedList::new()),
+        }
+    }
+
+    /// Wait for a notification.
+    ///
+    /// Each `Notify` value holds a single permit. If a permit is available from
+    /// an earlier call to [`notify()`], then `notified().await` will complete
+    /// immediately, consuming that permit. Otherwise, `notified().await` waits
+    /// for a permit to be made available by the next call to `notify()`
```
Comment (hawkw) [inline tokio/src/sync/notify.rs]: nit:
```suggestion
    /// for a permit to be made available by the next call to `notify()`.
```

File: tokio/src/sync/notify.rs [hunk 21]
```diff
@@ -0,0 +1,519 @@
...
+    pub async fn notified(&self) {
+        Notified {
+            notify: self,
+            state: State::Init,
+            waiter: linked_list::Entry::new(Waiter {
+                waker: None,
+                notified: false,
+            }),
+        }
+        .await
```
Comment (hawkw) [inline tokio/src/sync/notify.rs]: I'm assuming we do this rather than just returning the `Notified` future because we don't want to commit to making it public API?

File: tokio/src/sync/notify.rs [hunk 22]
```diff
@@ -0,0 +1,519 @@
...
+    }
+
+    /// Notifies a waiting task
+    ///
+    /// If a task is currently waiting, that task is notified. Otherwise, a
+    /// permit is stored in this `Notify` value and the **next** call to
+    /// `notified().await` will complete immediately consuming the permit made
```
Comment (hawkw) [inline tokio/src/sync/notify.rs]: nit/tioli: should we link to `notified` here?

File: tokio/src/sync/notify.rs [hunk 23]
```diff
@@ -0,0 +1,519 @@
...
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+// ===== impl Notified =====
+
+impl Notified<'_> {
+    fn project(
```
Comment (hawkw) [inline tokio/src/sync/notify.rs]: nit: might be worth commenting that we have to implement a manual pin projection for this type because `pin_project_lite` doesn't work with types implementing `Drop`?

File: tokio/src/macros/pin.rs (hunk 18 above)
Comment (carllerche) [reply tokio/src/macros/pin.rs]: failed merge!

File: tokio/src/sync/notify.rs (hunk 21 above)
Comment (carllerche) [reply tokio/src/sync/notify.rs]: yes... and docs.
